File: hash/string_appeal.py

def sum_total_appeal(string: str) -> int:  # LeetCode 2262.
    """
    Number of substrings ending at ith idx that contain a specific char
    = 1 + the last occurrence idx of the char before (i + 1)th idx.

    Total appeal of all substrings ending at ith idx
    = sum of number of substrings containing each character.
    """
    total_appeal, last_appeal = 0, 0
    chars_inclusions = dict()  # Number of substrings containing each char.
    for idx, char in enumerate(string):
        if char in chars_inclusions.keys():
            last_appeal -= chars_inclusions[char]
        last_appeal += 1 + idx

        total_appeal += last_appeal
        chars_inclusions.update({char: 1 + idx})

    return total_appeal

File: hash/string_appeal/string_appeal.cpp
#include <string>
#include <unordered_map>
using namespace std;

long long sumTotalAppeal(string s)
{ // LeetCode 2262.
    long long totalAppeal = 0, lastAppeal = 0;
    unordered_map<char, int> charsIndices; // Each char's last occurred idx.

    for (int idx = 0; idx < s.length(); idx++)
    {
        if (charsIndices.find(s[idx]) != charsIndices.end())
            lastAppeal -= 1 + charsIndices[s[idx]];

        charsIndices[s[idx]] = idx;

        lastAppeal += 1 + idx;
        totalAppeal += lastAppeal;
    }
    return totalAppeal;
}
